Make Weathervaning Accessible to Copter et al

Gone4Dirt opened the pull request:
This PR adds a number of weathervaning related features:

- Makes weathervaning functionality common to both plane and copter (useful for aerodynamically shaped bodies).
- Adds more options to control when weathervaning is activated or not, namely a velocity threshold.
- Adds a pitch component to the 'nose into wind' yaw rate calculation to keep a high yaw rate when the aircrafts tail is into wind.
- Adds the 'side into wind' option for copter tail sitters as per PR: #16292
- Adds autotests for weathervaning. 

These are feature requests are from two partners.

Weathervaning on copters has been tested in real life.

The 'side into wind' option has also been tested in real life and is working as desired.

I chose to add the weathervaning class to the AC_AttitudeControl library based on the fact that both Copters and VTOL planes have to be using the copter position controller to make use of weathervaning.  Happy to move it to its own library if that is preferred though.

andyp1per commented:
Please, please can we have an ifdef for this - very unlikely to be used on 1Mb boards and flash rapidly disappearing :(

muramura commented on ArduCopter/mode_auto.cpp:
```diff
@@ -833,6 +852,8 @@ void ModeAuto::wp_run()
     if (auto_yaw.mode() == AUTO_YAW_HOLD) {
         // roll & pitch from waypoint controller, yaw rate from pilot
         attitude_control->input_thrust_vector_rate_heading(wp_nav->get_thrust_vector(), target_yaw_rate);
+    } else if (auto_yaw.mode() == AUTO_YAW_RATE) {
```
By making it a SWITCH statement, I think it will be clear that it is a difference in mode processing.

muramura commented on libraries/AC_AttitudeControl/AC_WeatherVane.cpp:
```diff
@@ -0,0 +1,247 @@
+    // @DisplayName: Weathervaning min altitude
+    // @Description: Above this altitude weathervaning is permitted.  If terrain is enabled, AGL is used.  Set zero to ignore altitude.
+    // @Units: m
+    // @Range: 0 50
```
The ground also has a negative altitude.
In Japan, there is ground with negative altitude.

Gone4Dirt replied:
I have clarified the wording in the metadata.  This is actually a height not an altitude so that was my misuse of the word.  If a user wishes to fly close to the ground in locations other than home and have the weathervane activation dependant on that height, then they should enable terrain.

Gone4Dirt commented:
> Please, please can we have an ifdef for this - very unlikely to be used on 1Mb boards and flash rapidly disappearing :(

Done

tridge commented:
I love the idea of this, just needs to wait for after 4.1

rmackay9 commented:
I'm ok with this idea and it's nice to see it implemented as a yaw mode.  We might want to have @lthall have a peek at the algorithm if he's got time.

Gone4Dirt commented on libraries/AC_AttitudeControl/AC_WeatherVane.cpp:
```diff
@@ -0,0 +1,234 @@
+    AP_Param::setup_object_defaults(this, var_info);
+}
+
+float AC_WeatherVane::get_yaw_rate_cds(const int16_t roll_cdeg, const int16_t pitch_cdeg)
```
Sorry, I don't follow why degrees float is prefered here.  int16_t cdeg is how we recieve the roll and pitch from wp_nav so this makes it cleaner no?  The calculations within the function are done as floats and returned as such.

tridge commented on libraries/AC_AttitudeControl/AC_WeatherVane.cpp:
```diff
@@ -0,0 +1,235 @@
+    return true;
+}
+
+bool AC_WeatherVane::below_min_height(void)
```
this doesn't look like a match to the existing function, instead pass height above ground from Plane::relative_ground_altitude()

rmackay9 commented on libraries/AC_AttitudeControl/AC_WeatherVane.cpp:
```diff
@@ -0,0 +1,235 @@
+    // @User: Standard
+    AP_GROUPINFO("HGT_MIN", 4, AC_WeatherVane, _min_height, 2),
+
+    // @Param: VXY_MAX
```
I might call this param VEL_MAX or SPEED_MAX and then just clarify in the display name and/or descxription that only the horizontal speed is used

tridge replied:
SPEED_MAX would be good I think

tridge commented on libraries/AC_AttitudeControl/AC_WeatherVane.cpp:
```diff
@@ -0,0 +1,235 @@
+    first_activate_ms = 0;
+}
+
+bool AC_WeatherVane::should_weathervane(const int16_t pilot_yaw, const int16_t roll_cdeg, const int16_t pitch_cdeg)
```
the pilot_yaw input is a bit vehicle specific

Gone4Dirt replied:
pilot_yaw is to check if we have had a pilot's command to yaw ti make sure we aren't fighting the pilot.  The variable can be an angle, rate, or rcin.  We just need to check it is non-zero.  So the name seems appropriate to me, but I open to other suggestions.

rmackay9 commented on libraries/AC_AttitudeControl/AC_WeatherVane.cpp:
```diff
@@ -0,0 +1,235 @@
+    // @User: Standard
+    AP_GROUPINFO("VXY_MAX", 5, AC_WeatherVane, _max_vel_xy, 2.0),
+
+    // @Param: VZ_MAX
```
maybe rename this parameter to VELZ_MAX if you've got enough characters.

tridge commented on libraries/AC_AttitudeControl/AC_WeatherVane.cpp:
```diff
@@ -0,0 +1,235 @@
+    }
+
+    // Check if we meet the horizontal velocity thresholds to allow weathervaning
+    if ((_max_vel_xy.get() > 0) && (_inav.get_speed_xy()*0.01) > _max_vel_xy.get()) {
```
better to use AP::ahrs() instead of inav

tridge commented on libraries/AC_AttitudeControl/AC_WeatherVane.cpp:
```diff
@@ -0,0 +1,235 @@
+    }
+#endif
+
+    if (_inav.get_altitude() >= (float)_min_height.get()) {
```
caught in the inav cm trap!

tridge commented on libraries/AC_AttitudeControl/AC_WeatherVane.h:
```diff
@@ -0,0 +1,68 @@
+        AC_WeatherVane(const AP_InertialNav& inav);
+
+        // Do not allow copies
+        AC_WeatherVane(const AC_WeatherVane &other) = delete;
```
we've got a macro for this CLASS_NO_COPY()

tridge commented on libraries/AC_AttitudeControl/AC_WeatherVane.h:
```diff
@@ -0,0 +1,68 @@
+    private:
+
+        // References to other libraries
+        const AP_InertialNav& _inav;
```
remove inav, use ahrs

tridge commented on libraries/AC_AttitudeControl/AC_WeatherVane.h:
```diff
@@ -0,0 +1,68 @@
+        AP_Int8 _direction;
+        AP_Float _gain;
+        AP_Float _min_dz_ang_deg;
+        AP_Int32 _min_height;
```
change to AP_Float meters

tridge commented on libraries/AC_AttitudeControl/AC_WeatherVane.cpp:
```diff
@@ -0,0 +1,235 @@
+    // @Range: 0 50
+    // @Increment: 0.1
+    // @User: Standard
+    AP_GROUPINFO("VXY_MAX", 5, AC_WeatherVane, _max_vel_xy, 2.0),
```
needs to default to zero in quadplanes

tridge commented on libraries/AC_AttitudeControl/AC_WeatherVane.cpp:
```diff
@@ -0,0 +1,235 @@
+    // @Range: 0 5
+    // @Increment: 0.1
+    // @User: Standard
+    AP_GROUPINFO("VZ_MAX", 6, AC_WeatherVane, _max_vel_z, 1.0),
```
default to zero on quadplanes

tridge commented on libraries/AC_AttitudeControl/AC_WeatherVane.h:
```diff
@@ -0,0 +1,68 @@
+#if !(HAL_MINIMIZE_FEATURES && APM_BUILD_TYPE(APM_BUILD_ArduCopter))
```
make possible to disable with define in hwdef.dat

rmackay9 commented on ArduCopter/autoyaw.cpp:
```diff
@@ -264,3 +270,30 @@ float Mode::AutoYaw::rate_cds() const
     return 0.0f;
 }
+
+void Mode::AutoYaw::update_weathervane(const int16_t pilot_yaw, const int16_t roll_cdeg, const int16_t pitch_cdeg)
```
normally the order of inputs is roll, pitch, yaw...

Gone4Dirt replied:
changed the ordering to match rpy.  yaw can actually be an angle, rate, or rcin.  So long as it represents a pilot's request.

rmackay9 commented on ArduCopter/mode_guided.cpp:
```diff
@@ -739,6 +770,11 @@ void ModeGuided::posvelaccel_control_run()
     }
 
+    // if set and no pilot input for 2 sec weathervane copter into wind
+    if (allows_weathervaning()) {
```
I wonder if we want weathervaning available for all the guided mode controllers.  My feeling is that perhaps it should only be available for the wp control (e.g. what happens when the user clicks on "Fly Here" in the GCS).

Gone4Dirt replied:
Whilst I agree that it is less likely for weathervaning to be used in posvel its not impossible to image users wanting it.  I would much rather we leave it in so that it is more consistant across the guided controllers.  The differnce in guided controllers is already pretty confusing without adding to the differences.

rmackay9 commented on ArduCopter/autoyaw.cpp:
```diff
@@ -264,3 +270,30 @@ float Mode::AutoYaw::rate_cds() const
+            copter.g2.weathervane.set_relax(copter.ap.land_complete || copter.ap.land_complete_maybe);
+
+            // set yaw rate 
+            _yaw_rate_cds = copter.g2.weathervane.get_yaw_rate_cds(roll_cdeg, pitch_cdeg);
```
So the output is a rate instead of an angle?  .. and I guess this explains why we have a separate gain.  With the introduction of SCurves we support a simultaneous target angle and rate.

Gone4Dirt replied:
yes, output is rate.  We convert the roll angle to a yaw rate via the gain.

rmackay9 commented on ArduCopter/mode_auto.cpp:
```diff
@@ -846,6 +851,11 @@ void ModeAuto::wp_run()
 
+    // if set and no pilot input for 2 sec weathervane copter into wind
+    if (allows_weathervaning()) {
+        auto_yaw.update_weathervane(target_yaw_rate, wp_nav->get_roll(), wp_nav->get_pitch());
```
it's a bit of a drag that we have this yaw stuff creeping out into each controller and the input handling and decision of which yaw control method to use becomes fragmented.

Gone4Dirt replied:
Yes, agreed this could be better.  The 'process pilots yaw input' code block above this sets the precident.  I think we can be tidier and move all of the yaw handling into autoyaw and then we simply call an 'update_yaw' function instead.  But that should be a seperate PR.

Gone4Dirt commented:
Delay merge was added when we were rebasing off of master for 4.1.  Were past that point now so removing the delay merge label.

IamPete1 commented on ArduCopter/autoyaw.cpp:
```diff
@@ -56,6 +56,7 @@ void Mode::AutoYaw::set_mode(autopilot_yaw_mode yaw_mode)
     if (_mode == yaw_mode) {
         return;
     }
+    _last_mode = (autopilot_yaw_mode)_mode;
```
should probably just make `_mode`  a `autopilot_yaw_mode` type

Gone4Dirt replied:
Agreed. Done.  Added as seperate commit to keep it clear.

IamPete1 commented on ArduCopter/config.h:
```diff
@@ -332,6 +332,14 @@
+// Weathervane - allow vehicle to yaw into wind
+#ifndef WEATHERVANE_ENABLED
+# define WEATHERVANE_ENABLED !HAL_MINIMIZE_FEATURES
+#endif
```
Better to shove this in the library.

Gone4Dirt replied:
It's only enable/disabled in Copter.  Plane will have it as default.  It is a 'fixed' feature in quadplane as it were and I can't see why anyone would want to switch it off.  Does that sound correct to you?

IamPete1 commented on ArduCopter/mode_guided.cpp:
```diff
@@ -721,6 +735,22 @@ void ModeGuided::accel_control_run()
+        auto_yaw.update_weathervane(wp_nav->get_roll(), wp_nav->get_pitch(), target_yaw_rate, get_alt_above_ground_cm());
+    }
+
+    // landed with positive desired climb rate, initiate takeoff
```
fairly sure this is not meant to be here, just a rebase thing I guess.

Gone4Dirt replied:
yip, good catch, removed.

IamPete1 commented on libraries/AC_AttitudeControl/AC_WeatherVane.cpp:
```diff
@@ -0,0 +1,216 @@
+bool AC_WeatherVane::should_weathervane(const int16_t roll_cdeg, const int16_t pitch_cdeg, const int16_t pilot_yaw, const float hgt)
+{
+    // Check enabled
+    if (get_direction() == Direction::OFF){
```
```suggestion
    if (get_direction() == Direction::OFF) {
```

IamPete1 commented on libraries/AC_AttitudeControl/AC_WeatherVane.cpp:
```diff
@@ -0,0 +1,216 @@
+    // Check if we meet the horizontal velocity thresholds to allow weathervaning
+    const AP_AHRS &ahrs = AP::ahrs();
+    Vector3f vel_ned;
+    if (!ahrs.get_velocity_NED(vel_ned) || ((_max_vel_xy.get() > 0) && (sqrtf(vel_ned.x*vel_ned.x + vel_ned.y*vel_ned.y) > _max_vel_xy.get()))) {
```
if your only using it one you can also do this
```suggestion
    if (!AP::ahrs().get_velocity_NED(vel_ned) || ((_max_vel_xy.get() > 0) && (sqrtf(vel_ned.x*vel_ned.x + vel_ned.y*vel_ned.y) > _max_vel_xy.get()))) {
```
and skip the extra refence.

Gone4Dirt replied:
Done

IamPete1 commented on libraries/AC_AttitudeControl/AC_WeatherVane.cpp:
```diff
@@ -0,0 +1,216 @@
+#include <AP_Math/AP_Math.h>
+#include <GCS_MAVLink/GCS.h>
+#include <AP_Terrain/AP_Terrain.h>
+#include <AP_AHRS/AP_AHRS.h>
```
Fairly sure you don't need all of these.

Gone4Dirt replied:
yeah, some hangovers from old versions.  Have tidied.

IamPete1 commented on libraries/AC_AttitudeControl/AC_WeatherVane.h:
```diff
@@ -0,0 +1,54 @@
+#include <AP_Param/AP_Param.h>
+#include <AP_InertialNav/AP_InertialNav.h>
```
```suggestion
```

Gone4Dirt replied:
Removed

Gone4Dirt commented on ArduCopter/autoyaw.cpp:
```diff
@@ -56,40 +56,28 @@ void Mode::AutoYaw::set_mode(autopilot_yaw_mode yaw_mode)
-    case AUTO_YAW_LOOK_AT_NEXT_WP:
-        // wpnav will initialise heading when wpnav's set_destination method is called
-        break;
-
```
This diff looks a little weird here.  This is becuase I changed _mode to be of type autopilot_yaw_mode so needed to add the default case.  I then tidied up the cases and preserved the comments (below).  This has been done as a seperate commit to make it clearer.

tridge commented:
needs testing in its final form with quadplane (eg. Griffin or Alti model) and approval from Randy

Hwurzburg commented:
Also needs an entry in ap.py

Hwurzburg commented:
CustomBuild repo...the custom build server

robustini commented:
@Gone4Dirt Matt please, could you do remove the conflicts?
I have tests to do tomorrow and would like to try your PR with my tailsitter.

Gone4Dirt commented:
Thanks for testing @robustini your vid shows it working nicely.  Yeah, I have landed on the same gain of 3 for some large copters irl as well.  The low default gain is just to be conservative when someone first tries it on there vehicle.
Yes, there is the option to keep nose upwind.  The 'nose or tail into wind' was actually implemented with tail sitters in mind so that they can pitch either forward or back to lean into wind.

robustini commented:
> 
> 
> Thanks for testing @robustini your vid shows it working nicely. Yeah, I have landed on the same gain of 3 for some large copters irl as well. The low default gain is just to be conservative when someone first tries it on there vehicle. Yes, there is the option to keep nose upwind. The 'nose or tail into wind' was actually implemented with tail sitters in mind so that they can pitch either forward or back to lean into wind.

I do not find the description of the other parameters, I will look in the code.

robustini commented:
@Gone4Dirt I have done other tests and unfortunately when i reposition the tailsitter in QLoiter with Weathervaning enabled sometimes i have unwanted rotations on the yaw.
The automatism should be excluded when the tx sticks come out of the deadband, that is during a manual repositioning.
If any of the parameters can solve it I don't understand which one and how.

robustini commented:
> 
> 
> @Gone4Dirt to get this in sooner, you could split the copter changes out to a separate PR, and we could accept this in for plane now (with a fix for yawing while transitioning in tailsitters)

Surely it would be better that in the transition it is absolutely ignored, at the moment I solved it whit this settings: 

Q_WVANE_SPD_MAX,2
Q_WVANE_VELZ_MAX,2

robustini commented:
@Gone4Dirt Matt no news regarding @tridge's request?

robustini commented:
@Gone4Dirt please could you split the copter changes out to a separate PR so maybe we could do a merge for the plane with the tailsitter fixes? Thank you!

IamPete1 commented on ArduPlane/quadplane.cpp:
```diff
@@ -521,6 +513,10 @@ const AP_Param::ConversionInfo q_conversion_table[] = {
     { Parameters::k_param_quadplane, 22,  AP_PARAM_INT16, "Q_M_PWM_MIN" },
     { Parameters::k_param_quadplane, 23,  AP_PARAM_INT16, "Q_M_PWM_MAX" },
+
+    // PARAMETER_CONVERSION - Added: Sep-2021
```
```suggestion
    // PARAMETER_CONVERSION - Added: Jan-2022
```
;)

Gone4Dirt replied:
lol

IamPete1 commented on ArduPlane/quadplane.cpp:
```diff
@@ -685,6 +681,12 @@ bool QuadPlane::setup(void)
 
+    weathervane = new AC_WeatherVane();
+    if (!weathervane) {
+        AP_BoardConfig::config_error("Unable to allocate %s", "weathervane");
```
```suggestion
        AP_BoardConfig::allocation_error("weathervane");
```

IamPete1 commented on ArduPlane/quadplane.cpp:
```diff
@@ -3219,51 +3221,27 @@ float QuadPlane::get_weathervane_yaw_rate_cds(void)
+    // relax weathervane controller if we think we have landed
+    weathervane->set_relax(should_relax());
+
+    if (weathervane->should_weathervane(wp_nav->get_roll(), wp_nav->get_pitch(), plane.channel_rudder->get_control_in(), plane.relative_ground_altitude(plane.g.rangefinder_landing))) {
```
This is a existing issue, but we should probably be asking the attitude or position controller rather than WP nav. No need to do it in this PR.

IamPete1 commented on libraries/AC_AttitudeControl/AC_WeatherVane.cpp:
```diff
@@ -0,0 +1,228 @@
+// Constructor
+AC_WeatherVane::AC_WeatherVane(void)
+{
+    reset();
```
should not need to call this. It resets to 0's and false's. That will be the case after it is constructed anyway.

Gone4Dirt commented:
Rebased.

Feature requested: Have the ability to prevent weathervaning on take off for better control over the heading that the aircraft flys off to when transitioning into forward flight.  This has been implemented with the addition of allowing/preventing the weathervaning via an RC switch, and by extension by mavlink command.

Fixed the issue with it weathervaning immediately after the transition completes by adding the ability to reset the timer from quadplane.

robustini commented:
> Rebased.
> 
> Feature requested: Have the ability to prevent weathervaning on take off for better control over the heading that the aircraft flys off to when transitioning into forward flight. This has been implemented with the addition of allowing/preventing the weathervaning via an RC switch, and by extension by mavlink command.
> 
> Fixed the issue with it weathervaning immediately after the transition completes by adding the ability to reset the timer from quadplane.

@Gone4Dirt thanks Matt for the updates.
I am trying them but there is a problem, and I highlighted it on Discord here: https://discord.com/channels/674039678562861068/783141293445480470/927523972516569098

IamPete1 commented on libraries/RC_Channel/RC_Channel.cpp:
```diff
@@ -1185,6 +1187,8 @@ bool RC_Channel::do_aux_function(const aux_func_t ch_option, const AuxSwitchPos
     }
 
+
+
```
```suggestion
```
A little left over white space

IamPete1 commented on libraries/RC_Channel/RC_Channel.cpp:
```diff
@@ -46,6 +46,7 @@ extern const AP_HAL::HAL& hal;
 #include <AP_Mount/AP_Mount.h>
 #include <AP_VideoTX/AP_VideoTX.h>
 #include <AP_Torqeedo/AP_Torqeedo.h>
+#include <AC_AttitudeControl/AC_WeatherVane.h>
```
I don't think we need this?

IamPete1 commented on ArduPlane/quadplane.cpp:
```diff
@@ -3219,51 +3221,29 @@ float QuadPlane::get_weathervane_yaw_rate_cds(void)
-    if (should_relax()) {
-        output = 0;
+    // relax weathervane controller if we think we have landed
+    weathervane->set_relax(should_relax());
```
I wonder if we should just early return in this case, the relaxation sets the current output to zero, but the output is filtered, so the actual output might take some time to get down to 0. 

I would be tempted to add `should_relax()` to the above if that calls the reset and returns 0.

That means you may be able to remove the `set_relax` functionality all together.

hendjoshsr71 commented on ArduCopter/config.h:
```diff
@@ -326,6 +326,14 @@
+//////////////////////////////////////////////////////////////////////////////
+// Weathervane - allow vehicle to yaw into wind
+#ifndef WEATHERVANE_ENABLED
+# define WEATHERVANE_ENABLED !HAL_MINIMIZE_FEATURES
```
Is this what we are usually using instead?

```suggestion
# define WEATHERVANE_ENABLED BOARD_FLASH_SIZE <= 1024
```

Gone4Dirt replied:
Have we changed how we are doing this?  I though that the point of HAL_MINIMISE_FEATURES was exactly this:
https://github.com/ArduPilot/ardupilot/blob/427c08db2674ad4e02a7907a173b75a7f92bf5a8/libraries/AP_HAL/AP_HAL_Boards.h#L180

hendjoshsr71 replied:
Well we are trying to avoid adding more of those.  There are only like 9 boards `HAL_MINIMIZE_FEATURES 1` . So in the end them main defining feature to not compile something in is available flash.

rmackay9 commented on ArduCopter/mode.h:
```diff
@@ -274,6 +277,7 @@ class Mode {
     GCS_Copter &gcs();
     void set_throttle_takeoff(void);
     uint16_t get_pilot_speed_dn(void);
+    bool allows_weathervaning_auto(void) const;
```
let's move this somewhere else in the .h file because I don't think it's one of this "pass-through functions"..

Gone4Dirt replied:
done

hendjoshsr71 commented on Tools/autotest/arducopter.py:
```diff
@@ -3949,6 +3949,50 @@ def fly_payload_place_mission(self):
+    def weathervane_test(self):
+        # We test nose into wind code paths and yaw direction here and test side into wind
+        # yaw direction in QuadPlane tests to reduce repetition.
+        self.set_parameter("SIM_WIND_SPD", 10)
+        self.set_parameter("SIM_WIND_DIR", 100)
+        # allow weathervaning and arming from tx in guided
+        self.set_parameter("GUID_OPTIONS", 129)
+        # allow arming in auto, take off without raising the stick, and weathervaning
+        self.set_parameter("AUTO_OPTIONS", 131)
+        self.set_parameter("WVANE_ENABLE", 1)
+        self.set_parameter("WVANE_GAIN", 3)
+        self.set_parameter("WVANE_VELZ_MAX", 1)
+        self.set_parameter("WVANE_SPD_MAX", 2)
```
Peter switched to this for set_parameters as it is faster

```suggestion
        self.set_parameters({
            "SIM_WIND_SPD": 10,
            "SIM_WIND_DIR": 100,
            "GUID_OPTIONS": 129,    # allow weathervaning and arming from tx in guided
            "AUTO_OPTIONS": 131,    # allow arming in auto, take off without raising the stick, and weathervaning
            "WVANE_ENABLE": 1,
            "WVANE_GAIN"  : 3,
            "WVANE_VELZ_MAX": 1,
            "WVANE_SPD_MAX" : 2
        })
```

Gone4Dirt replied:
Done, Thanks

hendjoshsr71 commented on Tools/autotest/quadplane.py:
```diff
@@ -715,6 +715,28 @@ def test_pilot_yaw(self):
+    def weathervane_test(self):
+        # We test nose into wind code paths and yaw direction in copter autotest,
+        # so we shall test the side into wind yaw direction and plane code paths here.
+        self.set_parameter("SIM_WIND_SPD", 10)
+        self.set_parameter("SIM_WIND_DIR", 240)
+        # WVANE_ENABLE = 3 gives direction of side into wind
+        self.set_parameter("Q_WVANE_ENABLE", 3)
+        self.set_parameter("Q_WVANE_GAIN", 3)
+        self.set_parameter("STICK_MIXING", 0)
```
```suggestion
        self.set_parameters({
            "SIM_WIND_SPD": 10,
            "SIM_WIND_DIR": 240,
            "Q_WVANE_ENABLE": 3,    # WVANE_ENABLE = 3 gives direction of side into wind
            "Q_WVANE_GAIN": 3,
            "STICK_MIXING": 0,
        })
```

Gone4Dirt replied:
Done, thanks

IamPete1 commented on ArduPlane/quadplane.cpp:
```diff
@@ -3219,51 +3221,28 @@ float QuadPlane::get_weathervane_yaw_rate_cds(void)
+        return 0.0;
     }
 
-    float roll = wp_nav->get_roll() / 100.0f;
```
We have lost this divide by 100, hence the gain is much higher.

Gone4Dirt replied:
This isn't the issue with the gain.  The factor of 100 gets re-applied in the return of the function
https://github.com/ArduPilot/ardupilot/blob/f50f8c7a2a46a772312295e2ae03c0e78bd71418/ArduPlane/quadplane.cpp#L3279

IamPete1 commented on ArduPlane/quadplane.cpp:
```diff
@@ -3219,51 +3221,28 @@ float QuadPlane::get_weathervane_yaw_rate_cds(void)
-        weathervane.last_output = 0;
-        return 0;        
-    }
-    if (roll > 0) {
-        roll -= weathervane.min_roll;
-    } else {
-        roll += weathervane.min_roll;
-    }
```
We have also lost this logic, its abit odd, maybe it does not matter.

tridge commented on libraries/AC_AttitudeControl/AC_WeatherVane.cpp:
```diff
@@ -0,0 +1,219 @@
+            break;
+
+        case Direction::NOSE_IN:
+            output = (fabsf(roll_cdeg) + MAX(pitch_cdeg,0.0)) * _gain;
```
double check the gains match the old code

tridge commented on libraries/AC_AttitudeControl/AC_WeatherVane.cpp:
```diff
@@ -0,0 +1,219 @@
+
+    // Check if we meet the horizontal velocity thresholds to allow weathervaning
+    Vector3f vel_ned;
+    if (!AP::ahrs().get_velocity_NED(vel_ned) || ((_max_vel_xy.get() > 0) && (sqrtf(vel_ned.x*vel_ned.x + vel_ned.y*vel_ned.y) > _max_vel_xy.get()))) {
```
should be length_squared() and sq()

tridge commented on libraries/AC_AttitudeControl/AC_WeatherVane.cpp:
```diff
@@ -0,0 +1,219 @@
+    // @Range: 0 5
+    // @Increment: 0.1
+    // @User: Standard
+    AP_GROUPINFO("VELZ_MAX", 6, AC_WeatherVane, _max_vel_z, WVANE_PARAM_VELZ_MAX_DEFAULT),
```
need to default disable Z vel check on quadplane
